source/extensions/common/async_files/async_file_manager_thread_pool.cc

```diff
@@ -0,0 +1,253 @@
+      open_result = posix().open(path_.c_str(), O_TMPFILE | O_RDWR, S_IRUSR | S_IWUSR);
+      if (open_result.return_value_ == -1) {
+        return statusAfterFileError(open_result);
+      }
+      return std::make_shared<AsyncFileContextThreadPool>(manager_, open_result.return_value_);
+    }
+    // If O_TMPFILE didn't work, fall back to creating a named file and unlinking it.
+    // Use a fixed-size buffer because we're going to be using C file functions anyway, and it saves
+    // a heap allocation.
+    char filename[4096];
+    static const char file_suffix[] = "/buffer.XXXXXX";
+    if (path_.size() + sizeof(file_suffix) > sizeof(filename)) {
+      return absl::InvalidArgumentError(
+          "AsyncFileManagerThreadPool::createAnonymousFile: pathname too long for tmpfile");
+    }
+    // Using C-style functions here because `mkstemp` requires a writable
+    // char buffer, and we can't give it that with C++ strings.
+    snprintf(filename, sizeof(filename), "%s%s", path_.c_str(), file_suffix);
+    open_result = posix().mkstemp(filename);
+    if (open_result.return_value_ == -1) {
+      return statusAfterFileError(open_result);
+    }
+    if (posix().unlink(filename).return_value_ != 0) {
+      // Most likely the problem here is we can't unlink a file while it's open - since that's a
+      // prerequisite of the desired behavior of this function, and we don't want to accidentally
+      // fill a disk with named tmp files, if this happens we close the file, unlink it, and report
+      // an error.
+      posix().close(open_result.return_value_);
+      posix().unlink(filename);
+      return absl::UnimplementedError(
+          "AsyncFileManagerThreadPool::createAnonymousFile: not supported for "
+          "target filesystem (failed to unlink an open file)");
+    }
+    return std::make_shared<AsyncFileContextThreadPool>(manager_, open_result.return_value_);
+  }
+
+private:
+  const std::string path_;
+};
+
+class ActionOpenExistingFile : public ActionWithFileResult {
+public:
+  ActionOpenExistingFile(AsyncFileManagerThreadPool& manager, absl::string_view filename,
+                         AsyncFileManager::Mode mode,
+                         std::function<void(absl::StatusOr<AsyncFileHandle>)> on_complete)
+      : ActionWithFileResult(manager, on_complete), filename_(filename), mode_(mode) {}
+
+  absl::StatusOr<AsyncFileHandle> executeImpl() override {
+    auto open_result = posix().open(filename_.c_str(), openFlags());
+    if (open_result.return_value_ == -1) {
+      return statusAfterFileError(open_result);
+    }
+    return std::make_shared<AsyncFileContextThreadPool>(manager_, open_result.return_value_);
+  }
+
+private:
+  int openFlags() const {
+    switch (mode_) {
+    case AsyncFileManager::Mode::ReadOnly:
+      return O_RDONLY;
+    case AsyncFileManager::Mode::WriteOnly:
+      return O_WRONLY;
+    case AsyncFileManager::Mode::ReadWrite:
+    default:
```

Comment:
remove default so we have a compiler error if something else is added?

Reply:
Done.